### PR TITLE
[SPARK-33400][SQL] Normalize sameOrderExpressions in SortOrder to avoid unnecessary sort operations

### PR DESCRIPTION
**Pull request by prakharjain09**
### What changes were proposed in this pull request?
This pull request tries to normalize the SortOrder properly to prevent unnecessary sort operators. Currently the sameOrderExpressions are not normalized as part of AliasAwareOutputOrdering.

Example: consider this join of three tables:

      """
        |SELECT t2id, t3.id as t3id
        |FROM (
        |    SELECT t1.id as t1id, t2.id as t2id
        |    FROM t1, t2
        |    WHERE t1.id = t2.id
        |) t12, t3
        |WHERE t1id = t3.id
      """.

The plan for this looks like:

      *(8) Project [t2id#1059L, id#1004L AS t3id#1060L]
      +- *(8) SortMergeJoin [t2id#1059L], [id#1004L], Inner
         :- *(5) Sort [t2id#1059L ASC NULLS FIRST ], false, 0         <-----------------------------
         :  +- *(5) Project [id#1000L AS t2id#1059L]
         :     +- *(5) SortMergeJoin [id#996L], [id#1000L], Inner
         :        :- *(2) Sort [id#996L ASC NULLS FIRST ], false, 0
         :        :  +- Exchange hashpartitioning(id#996L, 5), true, [id=#1426]
         :        :     +- *(1) Range (0, 10, step=1, splits=2)
         :        +- *(4) Sort [id#1000L ASC NULLS FIRST ], false, 0
         :           +- Exchange hashpartitioning(id#1000L, 5), true, [id=#1432]
         :              +- *(3) Range (0, 20, step=1, splits=2)
         +- *(7) Sort [id#1004L ASC NULLS FIRST ], false, 0
            +- Exchange hashpartitioning(id#1004L, 5), true, [id=#1443]
               +- *(6) Range (0, 30, step=1, splits=2)

In this plan, the marked sort node could have been avoided as the data is already sorted on "t2.id" by the lower SortMergeJoin.

### Why are the changes needed?
To remove unneeded Sort operators.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
New UT added.

### CODE REVIEW

**Comment by maropu** [comment]
This is related to https://github.com/apache/spark/pull/30300 ?

**Comment by prakharjain09** [comment]
@maropu I have updated the pull request to cover one scenario which was not covered by #30300 . Please review the changes.

**Comment by prakharjain09** [comment]
cc - @cloud-fan @imback82

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
LGTM if the existing tests pass.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputExpression.scala [hunk 1]
```diff
@@ -65,7 +65,11 @@ trait AliasAwareOutputOrdering extends AliasAwareOutputExpression {
 
   final override def outputOrdering: Seq[SortOrder] = {
     if (hasAlias) {
-      orderingExpressions.map(normalizeExpression(_).asInstanceOf[SortOrder])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputExpression.scala:68]: shall we include `sameOrderExpressions` as children of `SortOrder`? cc @viirya @maropu

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputExpression.scala:68]: Ah, I see. It seems we can and the refactoring looks fine to me. Do you have time to work on it, @prakharjain09, in a separate PR?

**Review by maropu — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputExpression.scala (hunk 1 above)
Comment (prakharjain09) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputExpression.scala:68]: @cloud-fan @maropu Sure. I will raise a separate PR to make sameOrderExpressions as children of sortorder - once this goes in.

**Comment by prakharjain09** [comment]
@maropu @cloud-fan Thanks for reviewing the changes. Please merge the changes if there are no further comments.

I will work on a followup PR to make sameOrderExpressions child of SortOrder.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**